File: LeetCodeWP/src/main/java/com/hef/review/review0101sort/InsertSort.java
package com.hef.review.review0101sort;

/**
 * @Date 2021/7/21
 * @Author lifei
 */
public class InsertSort {

    public static void sort(Comparable[] a) {
        if (a==null || a.length==0) return;
        for (int i=1; i<a.length; i++) {
            Comparable v = a[i];
            int j = i-1;
            while (j>=0 && less(v, a[j])) {
                a[j+1] = a[j];
                j--;
            }
            a[j+1] = v;
        }
    }

    private static boolean less(Comparable o1, Comparable o2) {
        return o1.compareTo(o2)<0;
    }
}

File: js_wp/InsertSort.js

// 插入排序
function insertSort(a) {
    if (!a || a.length == 0) {
        return;
    }
    for (let i=1; i<a.length; i++) {
        let j = i-1;
        let t = a[i];
        for (; j>=0; j--) {
            if (less(t, a[j])) {
                a[j+1] = a[j]; 
            } else {
                break;
            }
        }
        a[j+1] = t;
    }
}

function less(v1, v2) {
    if (v1<v2) {
        return true;
    } else {
        return false;
    }
}

let a = [3, 6, 1, 2, 9, 0, 1, 2, 3]
// let a = [3, 6, 1, 2, 9, 0, 1, 2, 3, 3, 3, 1, 2, 0, 0, 9]
console.log(a);
insertSort(a);
console.log(a);